repos/keg-cli/src/utils/services/publishService.js

```diff
@@ -0,0 +1,335 @@
+    { cwd: location },
+    false
+  )
+}
+
+/**
+ * TODO: refactor this to use the Git library
+ * Runs a set of git commands to create a new branch, commit the changes, and push to a remote
+ * 
+ * @function
+ * @param {Object} repo - Repo object containing meta-data about the current repo
+ * @param {Object} publishArgs - Define the state or the repo being published
+ * @param {Object} updated - Any repos that have already been published, within the publishContext
+ * @param {Object} params - Options passed from the command line
+ * 
+ * @returns {array} - Passed in updated array, with the passed in repo added
+ */
+const gitBranchCommitUpdates = async (repo, publishArgs, updated, params) => {
+
+  const { newVersion, context, remote='origin' } = publishArgs
+  const { dryrun } = params
+
+  try {
+    logFormal(repo, `Running commit service`)
+    // Build a new branch for the version
+    publishArgs.step = [ 5, 'git-branch']
+    const newBranch = `${context}-${newVersion || 'build-&-publish'}`
+    publishArgs.newBranch = newBranch
+
+    // Create a new branch for the repo and version
+    publishArgs.step = [ 6, 'git-checkout']
+    await runGitCmd(`checkout -b ${newBranch}`, repo.location)
+    publishArgs.currentBranch = newBranch
+
+    // Add the build changes
+    publishArgs.step = [ 7, 'git-add']
+    await runGitCmd(`add --all`, repo.location)
+
+    // Commit the changes
+    publishArgs.step = [ 8, 'git-commit']
+    const message = `Updating ${context} to version ${newVersion}`
+    !dryrun && await runGitCmd(`commit -m \"${message}\"`, repo.location)
+
+    // Push the branch to github
+    publishArgs.step = [ 9, 'git-push']
+    !dryrun && await runGitCmd(`push ${remote} ${newBranch}`, repo.location)
+  }
+  catch(err){
+    Logger.error(`Error creating git branch`, err.stack)
+    await rollbackChanges(repo, publishArgs)
+  }
+
+  // Add the update to updated, so we know this repo was published
+  return updated.concat([ {...repo, ...publishArgs} ])
+
+}
+
+/**
+ * Runs a set of yarn commands to test, build and publish a repo
+ * @function
+ * @param {Object} repo - Repo object containing meta-data about the current repo
+ * @param {Object} publishContext - Defines how the repo should be published
+ * @param {Object} publishArgs - Define the state or the repo being published
+ * @param {Object} params - Options passed from the command line
+ * 
+ * @returns {Boolean} - True if the repo was published to npm
+ */
+const repoYarnCommands = async (repo, publishContext, publishArgs, params) => {
+
+  const {
+    test,
+    build,
+    publish,
+    access='public',
+  } = publishContext.tasks
+  const { newVersion } = publishArgs
+  const { dryrun } = params 
+
+  try {
+    // Callback when an error is thrown for a repo script
+    const scriptError = script => async () => {
+      Logger.error(`Error running script ${script}`)
+      await rollbackChanges(repo, publishArgs)
+      return false
+    }
```

Comment:
I believe this is because the `yarn` commands thats being executed is done on a different session. so when it throws in that session, we don't get it on our session. so we have to use this callback